### PR TITLE
Scale back [Serializable] CoreFX types

### PR DESCRIPTION
**Pull request by morganbr**
This is the first step toward #19119. The intent is to only put SerializableAttribute on types where:

1. The types can be correctly serialized between runtimes, versions, and operating systems. This was determined by looking at types that have runtime or OS dependencies or are highly subject to change.
2. The types are clearly important for users to create their own serializable types. This was determined by analyzing serialization usage data along with adding types that complete sets or are needed in order to serialize other types. 

In all, about 100 types will continue to be serializable by BinaryFormatter. Of course, it's possible we've excluded an important type. We have chosen to start with a conservative set, but to expand it as necessary based on user feedback.

This change removes SerializableAttribute from the CoreFX types that didn't make the list as well as removing testing for serializing those types. 

Future changes will include:
* Making similar sweeps through CoreCLR and CoreRT
* Changes to make serializing these types compatible where their serialized form doesn't match .NET Framework.

### CODE REVIEW

**Comment by morganbr** [comment]
CC @dotnet/corert-contrib @mconnew @shmao

File: src/Common/tests/System/Collections/IEnumerable.Generic.Serialization.Tests.cs [hunk 1]
```diff
@@ -15,6 +15,13 @@ public abstract partial class IEnumerable_Generic_Tests<T> : TestBase<T>
         public void IGenericSharedAPI_SerializeDeserialize(int count)
         {
             IEnumerable<T> expected = GenericIEnumerableFactory(count);
+
+            // Not all IEnumerables are intended to be Serializable
+            if(!expected.GetType().IsSerializable)
+            {
+                return;
+            }
```
Comment (stephentoub) [inline src/Common/tests/System/Collections/IEnumerable.Generic.Serialization.Tests.cs]: We already have:
https://github.com/morganbr/corefx/blob/eec6217de02d28f0060c16a0b0e34a530269eb6b/src/Common/tests/System/Collections/IEnumerable.NonGeneric.Tests.cs#L65
for the non-generic counterpart.  We should ideally keep them similar.  But this is ok for now.

File: src/Common/tests/System/Collections/IEnumerable.Generic.Serialization.Tests.cs [hunk 2]
```diff
@@ -15,6 +15,13 @@ public abstract partial class IEnumerable_Generic_Tests<T> : TestBase<T>
         public void IGenericSharedAPI_SerializeDeserialize(int count)
         {
             IEnumerable<T> expected = GenericIEnumerableFactory(count);
+
+            // Not all IEnumerables are intended to be Serializable
+            if(!expected.GetType().IsSerializable)
```
Comment (stephentoub) [inline src/Common/tests/System/Collections/IEnumerable.Generic.Serialization.Tests.cs]: Nit: missing space after the if

File: src/System.Data.Common/src/System/Data/SQLTypes/SQLChars.cs [hunk 3]
```diff
@@ -13,7 +13,6 @@
 
 namespace System.Data.SqlTypes
 {
-    [Serializable, XmlSchemaProvider("GetXsdType")]
```
Comment (stephentoub) [inline src/System.Data.Common/src/System/Data/SQLTypes/SQLChars.cs]: Why did you remove XmlSchemaProvider?

File: src/System.Data.Common/src/System/Data/SQLTypes/SQLBinary.cs [hunk 4]
```diff
@@ -11,7 +11,6 @@
 
 namespace System.Data.SqlTypes
 {
-    [Serializable]
```
Comment (stephentoub) [inline src/System.Data.Common/src/System/Data/SQLTypes/SQLBinary.cs]: Should we keep all of the SqlXx types?  Seems a bit odd to have only a subset of them.

File: src/System.Drawing.Primitives/ref/System.Drawing.Primitives.cs [hunk 5]
```diff
@@ -7,7 +7,6 @@
 
 namespace System.Drawing
 {
-    [System.SerializableAttribute]
```
Comment (stephentoub) [inline src/System.Drawing.Primitives/ref/System.Drawing.Primitives.cs]: Isn't Color one we decided to keep serializable?

File: src/System.Runtime.Extensions/src/System/Collections/Comparer.cs [hunk 6]
```diff
@@ -17,7 +17,6 @@
 
 namespace System.Collections
 {
-    [Serializable]
```
Comment (stephentoub) [inline src/System.Runtime.Extensions/src/System/Collections/Comparer.cs]: Are all of the non-generic collections still serializable after this is removed?

File: src/System.Runtime/src/System/LazyOfTTMetadata.cs [hunk 7]
```diff
@@ -8,7 +8,6 @@
 
 namespace System
 {
-    [Serializable]
```
Comment (stephentoub) [inline src/System.Runtime/src/System/LazyOfTTMetadata.cs]: Depending on where things land with delegate, it's likely `Lazy<T>` would also need to be removed.

**Comment by stephentoub** [comment]
> It should include:

We also need to make sure that everything that's [Serializable] has tests.  And I'd suggest ensuring that for the types in corefx before merging this change; otherwise we may find that something that's serializable fails to serialize due to a dependency on one of these types serialization support is being removed from.  Hopefully it'll all "just work" though.

More general, that check list should also include ensuring we have tests that validate we can deserialize on whatever platform the tests are running state that was serialized on all of the relevant platforms (desktop, mono, coreclr, etc.)

File: src/Common/tests/System/Collections/IEnumerable.Generic.Serialization.Tests.cs (hunk 2 above)
Comment (morganbr) [inline src/Common/tests/System/Collections/IEnumerable.Generic.Serialization.Tests.cs]: Fixed.

File: src/System.Data.Common/src/System/Data/SQLTypes/SQLChars.cs (hunk 3 above)
Comment (morganbr) [inline src/System.Data.Common/src/System/Data/SQLTypes/SQLChars.cs]: Accidentally :) Fixed now.

File: src/System.Drawing.Primitives/ref/System.Drawing.Primitives.cs (hunk 5 above)
Comment (morganbr) [inline src/System.Drawing.Primitives/ref/System.Drawing.Primitives.cs]: Yes, but this is a reference assembly, which isn't really supposed to have these attributes.

File: src/System.Data.Common/src/System/Data/SQLTypes/SQLBinary.cs (hunk 4 above)
Comment (morganbr) [inline src/System.Data.Common/src/System/Data/SQLTypes/SQLBinary.cs]: I kept the equivalents of all of the primitives and I'll add SqlGuid in the next iteration. The rest are less obviously useful and we have no data supporting serializing them.

File: src/System.Runtime.Extensions/src/System/Collections/Comparer.cs (hunk 6 above)
Comment (morganbr) [inline src/System.Runtime.Extensions/src/System/Collections/Comparer.cs]: There didn't seem to be a problem found by the tests, but I agree that's fishy. I'll enable this and its dependencies, CompareInfo and SortVersion.

File: src/System.Runtime/src/System/LazyOfTTMetadata.cs (hunk 7 above)
Comment (morganbr) [inline src/System.Runtime/src/System/LazyOfTTMetadata.cs]: Lazy's delegate is marked NonSerialized -- it just initializes Value when serializing.

**Comment by morganbr** [comment]
@stephentoub , agreed that's a good set of testing to add, but I'd prefer to do it in a subsequent change. All of the collections already have serialization tests, which covers most of what's in CoreFX. The rest (mostly Sql*) have been inspected and are unlikely to have problems.

**Comment by morganbr** [comment]
@jkotas , done. Let me know if I missed anything.

File: src/System.Drawing.Primitives/ref/System.Drawing.Primitives.cs (hunk 5 above)
Comment (stephentoub) [inline src/System.Drawing.Primitives/ref/System.Drawing.Primitives.cs]: Ah, ok.

File: src/System.Runtime/src/System/LazyOfTTMetadata.cs (hunk 7 above)
Comment (stephentoub) [inline src/System.Runtime/src/System/LazyOfTTMetadata.cs]: Yup, you're right.

**Comment by stephentoub** [comment]
@dotnet-bot test Innerloop netfx Windows_NT Release x64 Build and Test please (#19824)

**Comment by stephentoub** [comment]
@morganbr, can you address the merge conflict?

File: src/System.Net.Http/src/System/Net/Http/HttpRequestException.cs [hunk 8]
```diff
@@ -6,7 +6,6 @@
 
 namespace System.Net.Http
 {
-    [Serializable]
     [SuppressMessage("Microsoft.Serialization", "CA2229")]
```
Comment (davidsh) [inline src/System.Net.Http/src/System/Net/Http/HttpRequestException.cs]: Shouldn't the SuppressMessage line be removed also since it is dealing with Serialization?

Comment (morganbr) [inline src/System.Net.Http/src/System/Net/Http/HttpRequestException.cs]: I believe that suppression is still valid because it's a warning that the type implements ISerializable and doesn't have a serialization constructor.

**Comment by morganbr** [comment]
@dotnet-bot test Innerloop OSX10.12 Release x64 Build and Test

File: src/System.Net.Http/src/System/Net/Http/HttpRequestException.cs (hunk 8 above)
Comment (davidsh) [inline src/System.Net.Http/src/System/Net/Http/HttpRequestException.cs]: ?I believe that suppression is still valid because it's a warning that the type implements ISerializable

But if we're removing the Serializable attribute, then the type shouldn't be implementing ISerializable anymore.  So, why would there be a warning at all?

Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/HttpRequestException.cs]: > But if we're removing the Serializable attribute, then the type shouldn't be implementing ISerializable anymore

ISerializable is part of the type's signature/shape whereas [Serializable] is treated as an implementation detail.  We can remove [Serializable], but its ISerializable definition remains as its part of the type's APIs / standard.  The implementation will just throw.

**Comment by marek-safar** [comment]
@karelz @stephentoub is there any way to undo the damage caused by this massive breaking change? It breaks any existing code where user defines [Serializable] type with field of .NET type you removed serialization support from.

**Comment by karelz** [comment]
We had a lenghty discussion about where to draw the line - see #19119 summary.
Let's move the discussion there as all decision makers are on that thread.

@marek-safar can you please summarize your concerns and motivation? Is it general? Does Mono want to keep higher compat and this prevents you from using CoreFX source code (that could be solved by ifdefs).

**Comment by lmolkova** [comment]
I have intentionally added `[Serializable]` attribute on the `System.Diagnostics.Activity` class because it could be set/get from the LogicalCallContext on .NET 4.5.
This change broke the public preview2 of DiagnosticSource (on nuget).

According to @marek-safar feedback there are other types affected
@karelz  @stephentoub @morganbr 
Please let me know if this change be reverted or everyone needs to check and fix affected classes.

/cc @vancem @SergeyKanzhelev

**Comment by stephentoub** [comment]
>  have intentionally added [Serializable] attribute on the System.Diagnostics.Activity class because it could be set/get from the LogicalCallContext on .NET 4.5.

Is there a test for this?  Sounds like there should be.

> According to @marek-safar feedback there are other types affected

Marek's feedback is related to Mono's source sharing of corefx source.

> Please let me know if this change be reverted or everyone needs to check and fix affected classes?

We should address breaks individually.
cc: @danmosemsft

**Comment by lmolkova** [comment]
@stephentoub

> Is there a test for this? Sounds like there should be.
It only affects .NET45 and we do not have automated testing for it in corefx. We have tested it manually and agreed on the risk that it may break (https://github.com/dotnet/corefx/pull/17262#issuecomment-288204050)

**Comment by danmosemsft** [comment]
@weshaggard I see the thread in #17262. We obviously test .NET 4.x code already. What is special about this package -- because it's OOB? This kind of break exactly the concern @karelz had in #17262.

@lmolkova pleaes open a separate issue and if necessary we can re-spin 2.0.

**Comment by lmolkova** [comment]
@danmosemsft 
DiagnosticSoruce is not shipped in .NET framework, if this is what you are asking. We have testing for netfx which is I assume net 4.7 and there we do not use CallContext, so there is no issue with existing tests

**Comment by danmosemsft** [comment]
The code is .NET Framework (netfx) + your OOB package, right? So I'm wondering if our existing test mechanism  would work, iff the OOB package was simply installed.

**Comment by karelz** [comment]
@lmolkova 

> It only affects .NET45 and we do not have automated testing for it in corefx. We have tested it manually and agreed on the risk that it may break

That should not prevent us from having automated test on Desktop. It will run on 4.6.1 in regular runs.
The idea was that 4.5 support is very similar to 4.6.1 -- at the end of the day, the package owners (@vancem @brianrob) are responsible for ensuring it works on 4.5 (in this case via your team).

**Comment by karelz** [comment]
We test (currently) on .NET 4.6.1 -- that will change in future BTW.
@lmolkova AFAIK your code does not have any differences between 4.5 and 4.6.1.
Either way, I would recommend always to test the package manually prior to its shipping.

**Comment by lmolkova** [comment]
@karelz 

> The idea was that 4.5 support is very similar to 4.6.1 

This is exactly the case where the 4.5 is different from 4.6.1.
We use CallContext on 4.5 only and Serializable attribute is only important there.

@danmosemsft 
> The code is .NET Framework (netfx) + your OOB package, right? So I'm wondering if our existing test mechanism would work, iff the OOB package was simply installed.

Existing tests work and we run tests on netfx, but they do not touch code specific to NET 4.5 - exactly this case

**Comment by karelz** [comment]
In that case, it is exactly the problem class I predicted. Please work with @vancem @brianrob on a plan to test the package on 4.5 prior to each shipping. Especially the 4.5 specifics.

Note that you can create 4.5-only tests in CoreFX, they won't run automatically, but you can run them manually on machine with 4.5 to make your sign offs easier ...

**Comment by karelz** [comment]
cc @valenis
